### PR TITLE
bpo-37319: Deprecated support of non-integer arguments in random.randrange().

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue37319

### CODE REVIEW

**Comment by rhettinger** [comment]
I don't think this should be done.  It churns an API that has been stable for two decades.  It doesn't solve a real problem. It may break working code.  It makes the code look gross. It will complicate maintenance.

File: Lib/random.py [hunk 1]
```diff
@@ -67,7 +68,7 @@
 SG_MAGICCONST = 1.0 + _log(4.5)
 BPF = 53        # Number of bits in a float
 RECIP_BPF = 2**-BPF
-
+_ONE = 1
```
Comment (rhettinger) [inline Lib/random.py]: Why would you do this? This seems like bad coding style.  Nothing better communicates the constant one than the constant one.

Comment (serhiy-storchaka) [reply Lib/random.py]: To distinguish the default value for `step` from other values which are equal to 1 (e.g. 1.0, Fraction(1, 1), Decimal('1'), etc).

There are two issues with using just
```python
    if step is 1:
```
1. It is not guaranteed that 1 is a singleton. It is a CPython implementation detail.
2. For this reason this code emits a compiler warning.

**Comment by alimcmaster1** [comment]
Looks like this is good to close from the discussion on https://bugs.python.org/issue37319 ?

**Comment by serhiy-storchaka** [comment]
Reopened because @rhettinger just opened a duplicate issue.

According to the same microbenchmark as in #23064, this PR provides a 10% speed up:
```
$ ./python -m timeit -s 'from random import randrange' 'randrange(15)' # baseline
500000 loops, best of 5: 515 nsec per loop
$ ./python -m timeit -s 'from random import randrange' 'randrange(15)'  # new version
500000 loops, best of 5: 467 nsec per loop
```

File: Lib/random.py (hunk 1 above)
Comment (rhettinger) [reply Lib/random.py]: Please just use the constant 1 and use an equality test instead of identity test.

File: Lib/random.py [hunk 2]
```diff
@@ -297,28 +299,44 @@ def randrange(self, start, stop=None, step=1):
...
         # Non-unit step argument supplied.
-        istep = int(step)
-        if istep != step:
-            raise ValueError("non-integer step for randrange()")
+        try:
+            istep = _index(step)
+        except TypeError:
+            istep = int(step)
+            if istep != step:
+                raise ValueError("non-integer step for randrange()")
```
Comment (rhettinger) [inline Lib/random.py:345]: While we're at it, the exception type should be converted to *TypeError*.

Comment (serhiy-storchaka) [reply Lib/random.py:345]: Should we raise it with custom message or reraise the exception raised by `index()`?

**Comment by serhiy-storchaka** [comment]
@rhettinger, what are your problems with this PR? I addressed your comments, and would appreciate any other suggestions.

**Comment by rhettinger** [comment]
I merged in PR 23064.   If you want to make further modifications or improvements that would be welcome.

**Comment by serhiy-storchaka** [comment]
I merged with master. Please make review.

**Pull request merged**